File: src/test/java/no/nav/sosialhjelp/soknad/business/service/dialog/SistInnsendteSoknadServiceTest.java
package no.nav.sosialhjelp.soknad.business.service.dialog;

import no.nav.sosialhjelp.soknad.business.db.repositories.soknadmetadata.SoknadMetadataRepository;
import no.nav.sosialhjelp.soknad.business.domain.SoknadMetadata;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Collections;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class SistInnsendteSoknadServiceTest {

    @Mock
    private SoknadMetadataRepository soknadMetadataRepository;

    @InjectMocks
    private SistInnsendteSoknadService sistInnsendteSoknadService;

    private final String fnr = "12345";
    private final String fiksForsendelseId = "fiksId";
    private final String navEnhet = "1234";

    @Test
    void skalHenteSistInnsendteSoknadForBruker() {
        var innsendtDato = LocalDateTime.now().minusDays(1);
        var soknadMetadata = new SoknadMetadata();
        soknadMetadata.fnr = fnr;
        soknadMetadata.fiksForsendelseId = fiksForsendelseId;
        soknadMetadata.navEnhet = navEnhet;
        soknadMetadata.innsendtDato = innsendtDato;

        when(soknadMetadataRepository.hentAlleInnsendteSoknaderForBruker(any()))
                .thenReturn(Collections.singletonList(soknadMetadata));

        var dto = sistInnsendteSoknadService.hentSistInnsendteSoknad(fnr);

        assertThat(dto).isPresent();
        assertThat(dto.get().getDigisosId()).isEqualTo(fiksForsendelseId);
        assertThat(dto.get().getEnhetsnr()).isEqualTo(navEnhet);
        assertThat(dto.get().getInnsendtDato()).isEqualTo(innsendtDato.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME));
    }

    @Test
    void skalHenteSistInnsendteSoknadForBruker_siste() {
        var innsendtDatoNyest = LocalDateTime.now().minusDays(1);
        var innsendtDatoElst = LocalDateTime.now().minusDays(2);

        var soknadMetadata1 = new SoknadMetadata();
        soknadMetadata1.fnr = fnr;
        soknadMetadata1.fiksForsendelseId = fiksForsendelseId;
        soknadMetadata1.navEnhet = navEnhet;
        soknadMetadata1.innsendtDato = innsendtDatoNyest;

        var soknadMetadata2 = new SoknadMetadata();
        soknadMetadata2.fnr = fnr;
        soknadMetadata2.fiksForsendelseId = "fiksId2";
        soknadMetadata2.navEnhet = navEnhet;
        soknadMetadata2.innsendtDato = innsendtDatoElst;

        when(soknadMetadataRepository.hentAlleInnsendteSoknaderForBruker(any()))
                .thenReturn(List.of(soknadMetadata1, soknadMetadata2));

        var dto = sistInnsendteSoknadService.hentSistInnsendteSoknad(fnr);

        assertThat(dto).isPresent();
        assertThat(dto.get().getDigisosId()).isEqualTo(fiksForsendelseId);
        assertThat(dto.get().getEnhetsnr()).isEqualTo(navEnhet);
        assertThat(dto.get().getInnsendtDato()).isEqualTo(innsendtDatoNyest.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME));
    }

    @Test
    void skalReturnereEmptyOptionalVedNull() {
        when(soknadMetadataRepository.hentAlleInnsendteSoknaderForBruker(any()))
                .thenReturn(null);

        var dto = sistInnsendteSoknadService.hentSistInnsendteSoknad(fnr);

        assertThat(dto).isEmpty();
    }

    @Test
    void skalReturnereEmptyOptionalVedTomListe() {
        when(soknadMetadataRepository.hentAlleInnsendteSoknaderForBruker(any()))
                .thenReturn(Collections.emptyList());

        var dto = sistInnsendteSoknadService.hentSistInnsendteSoknad(fnr);

        assertThat(dto).isEmpty();
    }

}

File: src/test/kotlin/no/nav/sosialhjelp/soknad/api/dialog/SistInnsendteSoknadServiceTest.kt
package no.nav.sosialhjelp.soknad.api.dialog

import io.mockk.every
import io.mockk.mockk
import no.nav.sosialhjelp.soknad.db.repositories.soknadmetadata.SoknadMetadata
import no.nav.sosialhjelp.soknad.db.repositories.soknadmetadata.SoknadMetadataRepository
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

internal class SistInnsendteSoknadServiceTest {
    private val soknadMetadataRepository: SoknadMetadataRepository = mockk()
    private val sistInnsendteSoknadService = SistInnsendteSoknadService(soknadMetadataRepository)

    private val fnr = "12345"
    private val fiksForsendelseId = "fiksId"
    private val navEnhet = "1234"

    @Test
    fun skalHenteSistInnsendteSoknadForBruker() {
        val innsendtDato = LocalDateTime.now().minusDays(1)
        val soknadMetadata = SoknadMetadata(
            id = 0L,
            behandlingsId = "behandlingsId",
            fnr = fnr,
            fiksForsendelseId = fiksForsendelseId,
            navEnhet = navEnhet,
            opprettetDato = innsendtDato,
            sistEndretDato = innsendtDato,
            innsendtDato = innsendtDato
        )
        every { soknadMetadataRepository.hentAlleInnsendteSoknaderForBruker(any()) } returns listOf(soknadMetadata)

        val dto = sistInnsendteSoknadService.hentSistInnsendteSoknad(fnr)
        assertThat(dto?.ident).isEqualTo(fnr)
        assertThat(dto?.navEnhet).isEqualTo(navEnhet)
        assertThat(dto?.innsendtDato).isEqualTo(innsendtDato.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME))
    }

    @Test
    fun skalHenteSistInnsendteSoknadForBruker_siste() {
        val innsendtDatoNyest = LocalDateTime.now().minusDays(1)
        val innsendtDatoEldst = LocalDateTime.now().minusDays(2)

        val soknadMetadata1 = SoknadMetadata(
            id = 0L,
            behandlingsId = "behandlingsId",
            fnr = fnr,
            fiksForsendelseId = fiksForsendelseId,
            navEnhet = navEnhet,
            opprettetDato = innsendtDatoNyest,
            sistEndretDato = innsendtDatoNyest,
            innsendtDato = innsendtDatoNyest
        )

        val soknadMetadata2 = SoknadMetadata(
            id = 1L,
            behandlingsId = "behandlingsId2",
            fnr = fnr,
            fiksForsendelseId = "fiksId2",
            navEnhet = navEnhet,
            opprettetDato = innsendtDatoEldst,
            sistEndretDato = innsendtDatoEldst,
            innsendtDato = innsendtDatoEldst
        )
        every { soknadMetadataRepository.hentAlleInnsendteSoknaderForBruker(any()) } returns listOf(
            soknadMetadata1,
            soknadMetadata2
        )

        val dto = sistInnsendteSoknadService.hentSistInnsendteSoknad(fnr)
        assertThat(dto?.ident).isEqualTo(fnr)
        assertThat(dto?.navEnhet).isEqualTo(navEnhet)
        assertThat(dto?.innsendtDato).isEqualTo(innsendtDatoNyest.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME))
    }

    @Test
    fun skalReturnereNullVedTomListe() {
        every { soknadMetadataRepository.hentAlleInnsendteSoknaderForBruker(any()) } returns emptyList()

        val dto = sistInnsendteSoknadService.hentSistInnsendteSoknad(fnr)
        assertThat(dto).isNull()
    }
}
